// @iiif/helpers/dist/fetch.global.js
"use strict";
var IIIFHelpers = (() => {
  var __defProp = Object.defineProperty;
  var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
  var __getOwnPropNames = Object.getOwnPropertyNames;
  var __hasOwnProp = Object.prototype.hasOwnProperty;
  var __export = (target, all) => {
    for (var name in all)
      __defProp(target, name, { get: all[name], enumerable: true });
  };
  var __copyProps = (to, from, except, desc) => {
    if (from && typeof from === "object" || typeof from === "function") {
      for (let key of __getOwnPropNames(from))
        if (!__hasOwnProp.call(to, key) && key !== except)
          __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
    }
    return to;
  };
  var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

  // src/fetch.ts
  var fetch_exports = {};
  __export(fetch_exports, {
    fetch: () => fetchAndUpgrade
  });

  // node_modules/.pnpm/@iiif+parser@2.2.0/node_modules/@iiif/parser/dist/chunk-J657UVVW.js
  var e = "http://library.stanford.edu/iiif/image-api/compliance.html#level0";
  var i = "http://library.stanford.edu/iiif/image-api/compliance.html#level1";
  var t = "http://library.stanford.edu/iiif/image-api/compliance.html#level2";
  var o = "http://library.stanford.edu/iiif/image-api/conformance.html#level0";
  var r = "http://library.stanford.edu/iiif/image-api/conformance.html#level1";
  var a = "http://library.stanford.edu/iiif/image-api/conformance.html#level2";
  var _ = "http://library.stanford.edu/iiif/image-api/1.1/compliance.html#level0";
  var I = "http://library.stanford.edu/iiif/image-api/1.1/compliance.html#level1";
  var l = "http://library.stanford.edu/iiif/image-api/1.1/compliance.html#level2";
  var p = "http://library.stanford.edu/iiif/image-api/1.1/conformance.html#level0";
  var s = "http://library.stanford.edu/iiif/image-api/1.1/conformance.html#level1";
  var n = "http://library.stanford.edu/iiif/image-api/1.1/conformance.html#level2";
  var E = "http://iiif.io/api/image/1/level0.json";
  var c = "http://iiif.io/api/image/1/profiles/level0.json";
  var F = "http://iiif.io/api/image/1/level1.json";
  var m = "http://iiif.io/api/image/1/profiles/level1.json";
  var f = "http://iiif.io/api/image/1/level2.json";
  var x = "http://iiif.io/api/image/1/profiles/level2.json";
  var A = "http://iiif.io/api/image/2/level0.json";
  var L = "http://iiif.io/api/image/2/profiles/level0.json";
  var h = "http://iiif.io/api/image/2/level1.json";
  var y = "http://iiif.io/api/image/2/profiles/level1.json";
  var d = "http://iiif.io/api/image/2/level2.json";
  var g = "http://iiif.io/api/image/2/profiles/level2.json";
  var M = "level0";
  var O = "level1";
  var v = "level2";
  var u = "http://iiif.io/api/image/2/level0";
  var N = "http://iiif.io/api/image/2/level1";
  var G = "http://iiif.io/api/image/2/level2";
  var R = [G, t, a, l, n, f, x, d, g, v];
  var C = [...R, N, i, r, I, s, F, m, h, y, O];
  var B = [u, N, G, e, i, t, o, r, a, _, I, l, p, s, n, E, c, F, m, f, x, A, L, h, y, d, g, M, O, v];
  var P = B;

  // node_modules/.pnpm/@iiif+parser@2.2.0/node_modules/@iiif/parser/dist/chunk-NJNTZ6QT.js
  function r2(e3) {
    for (let n2 in e3)
      (typeof e3[n2] > "u" || e3[n2] === null) && delete e3[n2];
    return e3;
  }
  function i2(e3) {
    return Array.isArray(e3) ? e3 : e3 ? [e3] : [];
  }

  // node_modules/.pnpm/@iiif+parser@2.2.0/node_modules/@iiif/parser/dist/chunk-D22QKJZO.js
  var d2 = Object.defineProperty;
  var e2 = (b2, a2, c2) => a2 in b2 ? d2(b2, a2, { enumerable: true, configurable: true, writable: true, value: c2 }) : b2[a2] = c2;
  var f2 = (b2, a2, c2) => (e2(b2, typeof a2 != "symbol" ? a2 + "" : a2, c2), c2);

  // node_modules/.pnpm/@iiif+parser@2.2.0/node_modules/@iiif/parser/dist/chunk-RCT3CZAV.js
  var P2 = ["sc:Collection", "sc:Manifest", "sc:Canvas", "sc:AnnotationList", "oa:Annotation", "sc:Range", "sc:Layer", "sc:Sequence", "oa:Choice", "Service", "ContentResource"];
  function S(t4) {
    if (typeof t4 > "u" || t4 === null)
      throw new Error("Null or undefined is not a valid entity.");
    if (Array.isArray(t4))
      throw new Error("Array is not a valid entity");
    if (typeof t4 != "object")
      throw new Error(`${typeof t4} is not a valid entity`);
    if (typeof t4["@type"] == "string") {
      let e3 = P2.indexOf(t4["@type"]);
      if (e3 !== -1)
        return P2[e3];
    }
    if (t4.profile)
      return "Service";
    if (t4.format || t4["@type"])
      return "ContentResource";
    throw new Error("Resource type is not known");
  }
  var h2 = class t2 {
    constructor(e3, n2 = {}) {
      f2(this, "traversals");
      f2(this, "options");
      this.traversals = { collection: [], manifest: [], canvas: [], annotationList: [], sequence: [], annotation: [], contentResource: [], choice: [], range: [], service: [], layer: [], ...e3 }, this.options = { convertPropsToArray: true, mergeMemberProperties: true, allowUndefinedReturn: false, ...n2 };
    }
    static all(e3) {
      return new t2({ collection: [e3], manifest: [e3], canvas: [e3], annotationList: [e3], sequence: [e3], annotation: [e3], contentResource: [e3], choice: [e3], range: [e3], service: [e3], layer: [e3] });
    }
    traverseCollection(e3) {
      return this.traverseType(this.traverseDescriptive(this.traverseLinking(this.traverseCollectionItems(e3))), this.traversals.collection);
    }
    traverseCollectionItems(e3) {
      if (this.options.mergeMemberProperties) {
        let n2 = [...(e3.manifests || []).map((r3) => typeof r3 == "string" ? { "@id": r3, "@type": "sc:Manifest" } : r3), ...(e3.collections || []).map((r3) => typeof r3 == "string" ? { "@id": r3, "@type": "sc:Collection" } : r3), ...e3.members || []], i3 = [], a2 = n2.filter((r3) => i3.includes(r3["@id"]) ? false : (i3.push(r3["@id"]), true));
        delete e3.collections, delete e3.manifests, e3.members = a2;
      }
      return e3.manifests && (e3.manifests = e3.manifests.map((n2) => this.traverseManifest(typeof n2 == "string" ? { "@id": n2, "@type": "sc:Manifest" } : n2))), e3.collections && (e3.collections = e3.collections.map((n2) => this.traverseCollection(typeof n2 == "string" ? { "@id": n2, "@type": "sc:Collection" } : n2))), e3.members && (e3.members = e3.members.map((n2) => typeof n2 == "string" ? n2 : n2["@type"] === "sc:Collection" ? this.traverseCollection(n2) : n2["@type"] === "sc:Manifest" ? this.traverseManifest(n2) : this.traverseUnknown(n2))), e3;
    }
    traverseManifest(e3) {
      return this.traverseType(this.traverseDescriptive(this.traverseLinking(this.traverseManifestItems(e3))), this.traversals.manifest);
    }
    traverseManifestItems(e3) {
      return e3.sequences && (e3.sequences = e3.sequences.map((n2) => this.traverseSequence(n2))), e3.structures && (e3.structures = e3.structures.map((n2) => this.traverseRange(n2))), e3;
    }
    traverseSequence(e3) {
      return this.traverseType(this.traverseDescriptive(this.traverseLinking(this.traverseSequenceItems(e3))), this.traversals.sequence);
    }
    traverseSequenceItems(e3) {
      return e3.canvases && (e3.canvases = e3.canvases.map((n2) => this.traverseCanvas(n2))), e3;
    }
    traverseCanvas(e3) {
      return this.traverseType(this.traverseDescriptive(this.traverseLinking(this.traverseCanvasItems(e3))), this.traversals.canvas);
    }
    traverseCanvasItems(e3) {
      return e3.images && (e3.images = e3.images.map((n2) => this.traverseAnnotation(n2))), e3.otherContent && (e3.otherContent = e3.otherContent.map((n2) => this.traverseAnnotationList(n2))), e3;
    }
    traverseRange(e3) {
      return this.traverseType(this.traverseDescriptive(this.traverseLinking(this.traverseRangeItems(e3))), this.traversals.range);
    }
    traverseRangeItems(e3) {
      if (this.options.mergeMemberProperties) {
        let n2 = [...(e3.ranges || []).map((i3) => typeof i3 == "string" ? { "@id": i3, "@type": "sc:Range" } : i3), ...(e3.canvases || []).map((i3) => typeof i3 == "string" ? { "@id": i3, "@type": "sc:Canvas" } : i3), ...e3.members || []];
        delete e3.ranges, delete e3.canvases, e3.members = n2.length ? n2.map((i3) => this.traverseUnknown(i3)) : void 0;
      }
      return e3;
    }
    traverseAnnotationList(e3) {
      let n2 = typeof e3 == "string" ? { "@id": e3, "@type": "sc:AnnotationList" } : e3;
      return this.traverseType(this.traverseDescriptive(this.traverseAnnotationListItems(n2)), this.traversals.annotationList);
    }
    traverseAnnotationListItems(e3) {
      return e3.resources && (e3.resources = e3.resources.map((n2) => this.traverseAnnotation(n2))), e3;
    }
    traverseAnnotation(e3) {
      return this.traverseType(this.traverseDescriptive(this.traverseLinking(this.traverseAnnotationItems(e3))), this.traversals.annotation);
    }
    traverseAnnotationItems(e3) {
      return e3.resource && (Array.isArray(e3.resource) ? e3.resource = e3.resource.map((n2) => this.traverseContentResource(n2)) : e3.resource = this.traverseContentResource(e3.resource)), e3.on, e3;
    }
    traverseLayer(e3) {
      return this.traverseType(this.traverseLinking(this.traverseLayerItems(e3)), this.traversals.layer);
    }
    traverseLayerItems(e3) {
      return e3.otherContent && (e3.otherContent = e3.otherContent.map((n2) => this.traverseAnnotationList(n2))), e3;
    }
    traverseChoice(e3) {
      return this.traverseType(this.traverseChoiceItems(e3), this.traversals.choice);
    }
    traverseChoiceItems(e3) {
      return e3.default && e3.default !== "rdf:nil" && (e3.default = this.traverseContentResource(e3.default)), e3.item && e3.item !== "rdf:nil" && (e3.item = e3.item.map((n2) => this.traverseContentResource(n2))), e3;
    }
    traverseService(e3) {
      return this.traverseType(this.traverseLinking(e3), this.traversals.service);
    }
    traverseContentResource(e3) {
      return e3["@type"] === "oa:Choice" ? this.traverseChoice(e3) : this.traverseType(this.traverseDescriptive(this.traverseLinking(e3)), this.traversals.contentResource);
    }
    traverseUnknown(e3) {
      if (!e3["@type"] || typeof e3 == "string")
        return e3;
      switch (S(e3)) {
        case "sc:Collection":
          return this.traverseCollection(e3);
        case "sc:Manifest":
          return this.traverseManifest(e3);
        case "sc:Canvas":
          return this.traverseCanvas(e3);
        case "sc:Sequence":
          return this.traverseSequence(e3);
        case "sc:Range":
          return this.traverseRange(e3);
        case "oa:Annotation":
          return this.traverseAnnotation(e3);
        case "sc:AnnotationList":
          return this.traverseAnnotationList(e3);
        case "sc:Layer":
          return this.traverseLayer(e3);
        case "Service":
          return this.traverseService(e3);
        case "oa:Choice":
          return this.traverseChoice(e3);
        case "ContentResource":
          return this.traverseContentResource(e3);
      }
      return e3.profile ? this.traverseService(e3) : e3;
    }
    traverseImageResource(e3) {
      let n2 = Array.isArray(e3), i3 = Array.isArray(e3) ? e3 : [e3], a2 = [];
      for (let r3 of i3)
        typeof r3 == "string" ? a2.push(this.traverseContentResource({ "@id": r3, "@type": "dctypes:Image" })) : a2.push(this.traverseContentResource(r3));
      return !n2 && !this.options.convertPropsToArray ? a2[0] : a2;
    }
    traverseDescriptive(e3) {
      return e3.thumbnail && (e3.thumbnail = this.traverseImageResource(e3.thumbnail)), e3.logo && (e3.logo = this.traverseImageResource(e3.logo)), e3;
    }
    traverseOneOrMoreServices(e3) {
      let n2 = Array.isArray(e3), i3 = Array.isArray(e3) ? e3 : [e3], a2 = [];
      for (let r3 of i3)
        a2.push(this.traverseService(r3));
      return !n2 && !this.options.convertPropsToArray ? a2[0] : a2;
    }
    traverseLinking(e3) {
      return e3.related && (e3.related = this.traverseOneOrManyType(e3.related, this.traversals.contentResource)), e3.rendering && (e3.rendering = this.traverseOneOrManyType(e3.rendering, this.traversals.contentResource)), e3.service && (e3.service = this.traverseOneOrMoreServices(e3.service)), e3.seeAlso && (e3.seeAlso = this.traverseOneOrManyType(e3.seeAlso, this.traversals.contentResource)), e3.within && (typeof e3.within == "string" || (e3.within = this.traverseOneOrManyType(e3.within, this.traversals.contentResource))), e3.startCanvas && (typeof e3.startCanvas == "string" ? e3.startCanvas = this.traverseType({ "@id": e3.startCanvas, "@type": "sc:Canvas" }, this.traversals.canvas) : e3.startCanvas && this.traverseType(e3.startCanvas, this.traversals.canvas)), e3.contentLayer && (typeof e3.contentLayer == "string" ? e3.contentLayer = this.traverseLayer({ "@id": e3.contentLayer, "@type": "sc:Layer" }) : e3.contentLayer = this.traverseLayer(e3.contentLayer)), e3;
    }
    traverseOneOrManyType(e3, n2) {
      if (!Array.isArray(e3))
        if (this.options.convertPropsToArray)
          e3 = [e3];
        else
          return this.traverseType(e3, n2);
      return e3.map((i3) => this.traverseType(i3, n2));
    }
    traverseType(e3, n2) {
      return n2.reduce((i3, a2) => {
        let r3 = a2(i3);
        return typeof r3 > "u" && !this.options.allowUndefinedReturn ? i3 : r3;
      }, e3);
    }
  };
  var O2 = "http://library.stanford.edu/iiif/image-api/compliance.html#level1";
  var w = "http://library.stanford.edu/iiif/image-api/compliance.html#level2";
  var b = "http://library.stanford.edu/iiif/image-api/conformance.html#level1";
  var F2 = "http://library.stanford.edu/iiif/image-api/conformance.html#level2";
  var D = "http://library.stanford.edu/iiif/image-api/1.1/compliance.html#level1";
  var N2 = "http://library.stanford.edu/iiif/image-api/1.1/compliance.html#level2";
  var k = "http://library.stanford.edu/iiif/image-api/1.1/conformance.html#level1";
  var G2 = "http://library.stanford.edu/iiif/image-api/1.1/conformance.html#level2";
  var j = "http://iiif.io/api/image/1/level1.json";
  var q = "http://iiif.io/api/image/1/profiles/level1.json";
  var V = "http://iiif.io/api/image/1/level2.json";
  var U = "http://iiif.io/api/image/1/profiles/level2.json";
  var $ = "http://iiif.io/api/image/2/level1.json";
  var W = "http://iiif.io/api/image/2/profiles/level1.json";
  var B2 = "http://iiif.io/api/image/2/level2.json";
  var H = "http://iiif.io/api/image/2/profiles/level2.json";
  var J = "level1";
  var z = "level2";
  var K = "http://iiif.io/api/image/2/level1";
  var Q = "http://iiif.io/api/image/2/level2";
  var _2 = [K, Q, O2, w, b, F2, D, N2, k, G2, j, q, V, U, $, W, B2, H, J, z];
  var y2 = { attributionLabel: "Attribution", lang: "none", providerId: "http://example.org/provider", providerName: "" };
  function X(t4) {
    if (typeof t4 == "string")
      return [t4];
    if (!t4)
      return [];
    let e3 = Array.isArray(t4) ? t4 : [t4], n2 = [];
    for (let i3 of e3) {
      if (typeof i3 == "string") {
        n2.push(i3);
        continue;
      }
      n2.push({ "@language": i3["@language"] || i3.language, "@value": i3["@value"] || i3.value });
    }
    return n2;
  }
  function u2(t4, e3 = "none") {
    if (!t4)
      return { none: [""] };
    let n2 = X(t4), i3 = {};
    for (let a2 of n2) {
      if (typeof a2 == "string") {
        i3[e3] = i3[e3] ? i3[e3] : [], i3[e3].push(a2 || "");
        continue;
      }
      if (!a2["@language"]) {
        i3[e3] = i3[e3] ? i3[e3] : [], i3[e3].push(a2["@value"] || "");
        continue;
      }
      let r3 = a2["@language"];
      i3[r3] = i3[r3] ? i3[r3] : [], i3[r3].push(a2["@value"] || "");
    }
    return Object.keys(i3).length === 0 ? { none: [""] } : i3;
  }
  function L2(t4) {
    if (Array.isArray(t4))
      return L2(t4.find((e3) => typeof e3 == "string"));
    if (R.indexOf(t4) !== -1)
      return "level2";
    if (_2.indexOf(t4) !== -1)
      return "level1";
    if (P.indexOf(t4) !== -1)
      return "level0";
    if (typeof t4 == "string")
      return t4;
  }
  function Y(t4) {
    let e3 = Array.isArray(t4) ? t4 : [t4];
    for (let n2 of e3)
      switch (n2) {
        case "http://iiif.io/api/image/2/context.json":
        case "http://library.stanford.edu/iiif/image-api/1.1/compliance.html#level2":
          return "ImageService2";
        case "http://iiif.io/api/image/1/context.json":
        case "http://library.stanford.edu/iiif/image-api/1.1/context.json":
          return "ImageService1";
        case "http://iiif.io/api/annex/openannotation/context.json":
          return "ImageApiSelector";
      }
  }
  function Z(t4) {
    switch (t4) {
      case "http://iiif.io/api/image/2/level0.json":
      case "http://iiif.io/api/image/2/level1.json":
      case "http://iiif.io/api/image/2/level2.json":
        return "ImageService2";
      case "http://iiif.io/api/auth/1/kiosk":
      case "http://iiif.io/api/auth/1/login":
      case "http://iiif.io/api/auth/1/clickthrough":
      case "http://iiif.io/api/auth/1/external":
      case "http://iiif.io/api/auth/0/kiosk":
      case "http://iiif.io/api/auth/0/login":
      case "http://iiif.io/api/auth/0/clickthrough":
      case "http://iiif.io/api/auth/0/external":
        return "AuthCookieService1";
      case "http://iiif.io/api/auth/1/token":
      case "http://iiif.io/api/auth/0/token":
        return "AuthTokenService1";
      case "http://iiif.io/api/auth/1/logout":
      case "http://iiif.io/api/auth/0/logout":
        return "AuthLogoutService1";
      case "http://iiif.io/api/search/1/search":
      case "http://iiif.io/api/search/0/search":
        return "SearchService1";
      case "http://iiif.io/api/search/1/autocomplete":
      case "http://iiif.io/api/search/0/autocomplete":
        return "AutoCompleteService1";
    }
  }
  function R2(t4) {
    for (let e3 of ["sc", "oa", "dcterms", "dctypes", "iiif"])
      if (t4.startsWith(`${e3}:`))
        return t4.slice(e3.length + 1);
    return t4;
  }
  var ee = ["Collection", "Manifest", "Annotation", "AnnotationPage", "Range", "Service"];
  function m2(t4) {
    let e3 = t4["@id"] || t4.id, n2 = t4["@type"] || t4.type, i3 = t4.profile || void 0, a2 = t4["@context"] || void 0;
    if (i3) {
      let r3 = Z(i3);
      if (r3)
        return r3;
    }
    if (a2) {
      let r3 = Y(a2);
      if (r3)
        return r3;
    }
    if (n2) {
      if (Array.isArray(n2)) {
        if (n2.indexOf("oa:CssStylesheet") !== -1)
          return "CssStylesheet";
        if (n2.indexOf("cnt:ContentAsText") !== -1)
          return "TextualBody";
        n2 = n2[0];
      }
      for (let r3 of ["sc", "oa", "dcterms", "dctypes", "iiif"])
        if (n2.startsWith(`${r3}:`)) {
          n2 = n2.slice(r3.length + 1);
          break;
        }
      switch (n2) {
        case "Layer":
          return "AnnotationCollection";
        case "AnnotationList":
          return "AnnotationPage";
        case "cnt:ContentAsText":
          return "TextualBody";
      }
    }
    if (n2 && ee.indexOf(n2) !== -1)
      return n2;
    if (t4.format) {
      if (t4.format.startsWith("image/"))
        return "Image";
      if (t4.format.startsWith("text/") || t4.format === "application/pdf")
        return "Text";
      if (t4.format.startsWith("application/"))
        return "Dataset";
    }
    return e3 && (e3.endsWith(".jpg") || e3.endsWith(".png") || e3.endsWith(".jpeg")) ? "Image" : n2 || "unknown";
  }
  var te = /http(s)?:\/\/(creativecommons.org|rightsstatements.org)[^"'\\<\n]+/gm;
  function ne(t4) {
    let e3 = t4.match(te);
    return e3 ? e3[0] : t4;
  }
  function ie(t4, e3 = "Rights/License", n2 = "none") {
    let i3 = null, a2 = [], r3 = Array.isArray(t4) ? t4 : [t4];
    for (let s2 of r3) {
      let c2 = s2 ? ne(s2) : void 0;
      if (c2 && (c2.indexOf("creativecommons.org") !== -1 || c2.indexOf("rightsstatements.org") !== -1)) {
        c2.startsWith("https://") ? i3 = `http://${c2.slice(8)}` : i3 = c2;
        continue;
      }
      c2 && a2.push({ label: { [n2]: [e3] }, value: { [n2]: [c2] } });
    }
    return [i3, a2];
  }
  var re = ["http://iiif.io/api/presentation/2/context.json", "http://iiif.io/api/image/2/context.json", "http://iiif.io/api/image/1/context.json", "http://library.stanford.edu/iiif/image-api/1.1/context.json", "http://iiif.io/api/search/1/context.json", "http://iiif.io/api/search/0/context.json", "http://iiif.io/api/auth/1/context.json", "http://iiif.io/api/auth/0/context.json", "http://iiif.io/api/annex/openannotation/context.json"];
  function ae(t4) {
    if (t4) {
      let e3 = Array.isArray(t4) ? t4 : [t4], n2 = [];
      for (let i3 of e3)
        i3 === "http://iiif.io/api/presentation/2/context.json" && n2.push("http://iiif.io/api/presentation/3/context.json"), re.indexOf(i3) === -1 && n2.push(i3);
      if (e3.length)
        return n2.length === 1 ? n2[0] : n2;
    }
  }
  function se(t4) {
    return t4 ? t4.map((e3) => ({ label: u2(e3.label), value: u2(e3.value) })) : [];
  }
  var x2 = 0;
  function E2(t4, e3) {
    let n2 = encodeURI(t4.id || t4["@id"] || "").trim();
    return n2 && e3 ? `${n2}/${e3}` : n2 || (x2++, `http://example.org/${t4["@type"]}${e3 ? `/${e3}` : ""}/${x2}`);
  }
  function p2(t4) {
    let e3 = [...t4.behavior || []];
    t4.viewingHint && e3.push(t4.viewingHint);
    let n2;
    return Array.isArray(t4.motivation) ? n2 = t4.motivation.map(R2) : t4.motivation && (n2 = R2(t4.motivation)), { "@context": t4["@context"] ? ae(t4["@context"]) : void 0, id: (t4["@id"] || E2(t4)).trim(), type: m2(t4), behavior: e3.length ? e3 : void 0, height: t4.height ? t4.height : void 0, width: t4.width ? t4.width : void 0, motivation: n2, viewingDirection: t4.viewingDirection, profile: t4.profile, format: t4.format ? t4.format : void 0, duration: void 0, timeMode: void 0 };
  }
  function f3(t4) {
    let [e3, n2] = ie(t4.license), i3 = [...t4.metadata ? se(t4.metadata) : [], ...n2];
    return { rights: e3, metadata: i3.length ? i3 : void 0, label: t4.label ? u2(t4.label) : void 0, requiredStatement: t4.attribution ? { label: u2(y2.attributionLabel), value: u2(t4.attribution) } : void 0, navDate: t4.navDate, summary: t4.description ? u2(t4.description) : void 0, thumbnail: oe(t4.thumbnail) };
  }
  function oe(t4) {
    return t4 && (Array.isArray(t4) ? t4 : [t4]).map((n2) => typeof n2 == "string" ? { id: n2, type: "Image" } : (n2.type === "unknown" && (n2.type = "Image"), n2));
  }
  function pe(t4) {
    if (!t4.within)
      return;
    let e3 = Array.isArray(t4.within) ? t4.within : [t4.within], n2 = [];
    for (let i3 of e3)
      if (typeof i3 == "string") {
        if (i3)
          switch (t4["@type"]) {
            case "sc:Manifest":
              n2.push({ id: i3, type: "Collection" });
              break;
          }
      } else
        i3["@id"] && n2.push({ id: i3["@id"], type: m2(i3) });
    return n2.length ? n2 : void 0;
  }
  function l2(t4) {
    let e3 = t4.related ? Array.isArray(t4.related) ? t4.related : [t4.related] : [], n2 = t4.contentLayer;
    return { provider: t4.logo || e3.length ? [{ id: y2.providerId, type: "Agent", homepage: e3.length ? [e3[0]] : void 0, logo: t4.logo ? Array.isArray(t4.logo) ? t4.logo : [t4.logo] : void 0, label: u2(y2.providerName) }] : void 0, partOf: pe(t4), rendering: t4.rendering, seeAlso: t4.seeAlso, start: t4.startCanvas, service: t4.service ? i2(t4.service) : void 0, supplementary: n2 ? [n2] : void 0 };
  }
  function fe(t4) {
    return { chars: t4.chars, format: t4.format ? t4.format : void 0, language: t4.language };
  }
  function d3(t4, e3) {
    return t4 ? typeof t4 == "string" ? { id: t4, type: e3 } : typeof t4?.["@id"] == "string" ? { id: t4["@id"], type: e3 } : typeof t4.id == "string" ? { id: t4.id, type: e3 } : null : null;
  }
  function ce(t4) {
    let e3 = {};
    if (t4.first) {
      let n2 = d3(t4.first, "Collection");
      n2 && (e3.first = n2);
    }
    if ((t4.total || t4.total === 0) && (e3.total = t4.total), t4.prev) {
      let n2 = d3(t4.prev, "Collection");
      n2 && (e3.prev = n2);
    }
    if (t4.next) {
      let n2 = d3(t4.next, "Collection");
      n2 && (e3.next = n2);
    }
    return e3;
  }
  function le(t4) {
    let e3 = [];
    for (let n2 of t4) {
      let i3 = { ...n2 };
      i3.items && i3.items.length === 0 && delete i3.items, e3.push(i3);
    }
    return e3;
  }
  function ue(t4) {
    return r2({ ...p2(t4), ...f3(t4), ...l2(t4), ...ce(t4), items: le(t4.members) });
  }
  function he(t4) {
    let e3 = [], n2 = [], i3, a2;
    for (let s2 of t4.sequences || [])
      s2.canvases.length && e3.push(...s2.canvases), s2.behavior && n2.push(...s2.behavior), s2.viewingDirection && (a2 = s2.viewingDirection), s2.startCanvas && (i3 = s2.startCanvas);
    let r3 = p2(t4);
    return n2.length && (r3.behavior ? r3.behavior.push(...n2) : r3.behavior = n2), r2({ ...r3, ...f3(t4), ...l2(t4), viewingDirection: a2, start: i3, items: e3, structures: ve(t4.structures) });
  }
  function ve(t4) {
    if (!t4)
      return t4;
    let e3 = /* @__PURE__ */ new Map();
    for (let i3 of t4)
      e3.set(i3.id, i3);
    let n2 = [];
    for (let i3 of t4)
      if (i3.items) {
        let a2 = i3.items.map((r3) => typeof r3 == "string" ? (n2.push(r3), e3.get(r3) || r3) : r3 && r3.id ? (n2.push(r3.id), e3.get(r3.id) || r3) : r3);
        i3.items = a2;
      }
    return t4.filter((i3) => n2.indexOf(i3.id) === -1);
  }
  function de(t4) {
    return r2({ ...p2(t4), ...f3(t4), ...l2(t4), annotations: t4.otherContent && t4.otherContent.length ? t4.otherContent : void 0, items: t4.images && t4.images.length ? [{ id: E2(t4, "annotation-page"), type: "AnnotationPage", items: t4.images }] : void 0 });
  }
  function ye(t4) {
    return r2({ ...p2(t4), ...f3(t4), ...l2(t4), items: t4.resources && t4.resources.length ? t4.resources : void 0 });
  }
  function ge(t4) {
    return !t4.canvases || t4.canvases.length === 0 ? { canvases: [], behavior: [] } : { canvases: t4.canvases, behavior: t4.viewingHint ? [t4.viewingHint] : [], viewingDirection: t4.viewingDirection, startCanvas: t4.startCanvas };
  }
  function me(t4) {
    function e3(n2) {
      if (Array.isArray(n2)) {
        if (n2.length > 1)
          return { type: "List", items: n2.map(e3) };
        n2 = n2[0];
      }
      if (typeof n2 == "string")
        return encodeURI(n2).trim();
      if ("@type" in n2) {
        let i3;
        if (typeof n2.full == "string")
          i3 = n2.full;
        else if (n2.full["@type"] === "dctypes:Image")
          i3 = { id: n2.full["@id"], type: "Image" };
        else if (n2.full["@type"] === "sc:Canvas")
          i3 = { id: n2.full["@id"], type: "Canvas" };
        else
          throw new Error(`Unsupported source type on annotation: ${n2.full["@type"]}`);
        return { type: "SpecificResource", source: i3, selector: g2(n2.selector) };
      } else
        return encodeURI(n2["@id"]).trim();
    }
    return r2({ ...p2(t4), ...f3(t4), ...l2(t4), target: e3(t4.on), body: Array.isArray(t4.resource) ? t4.resource.map(T) : T(t4.resource) });
  }
  function T(t4) {
    return t4.type === "Choice" ? t4 : M2(t4);
  }
  function M2(t4) {
    let e3 = t4;
    return r2({ ...p2(e3), ...f3(e3), ...l2(e3), ...fe(e3) });
  }
  function Ce(t4) {
    let e3 = [];
    return t4.default && t4.default !== "rdf:nil" && e3.push(t4.default), t4.item && t4.item !== "rdf:nil" && e3.push(...t4.item), r2({ ...p2(t4), ...f3(t4), items: e3 });
  }
  function Ae(t4) {
    return r2({ ...p2(t4), ...f3(t4), ...l2(t4), items: t4.members });
  }
  function Ie(t4) {
    let { "@id": e3, "@type": n2, "@context": i3, profile: a2, ...r3 } = t4, s2 = {};
    return e3 && (s2["@id"] = e3), s2["@type"] = m2(t4), s2["@type"] === "unknown" && (i3 && i3.length && (s2["@context"] = i3), s2["@type"] = "Service"), a2 && (s2.profile = L2(a2)), r2({ ...s2, ...r3 });
  }
  function Pe(t4) {
    return r2({ ...p2(t4), ...f3(t4), ...l2(t4) });
  }
  var _e = new h2({ collection: [ue], manifest: [he], canvas: [de], annotationList: [ye], sequence: [ge], annotation: [me], contentResource: [M2], choice: [Ce], range: [Ae], service: [Ie], layer: [Pe] });
  function be(t4) {
    return t4 && t4["@context"] && (t4["@context"] === "http://iiif.io/api/presentation/2/context.json" || t4["@context"].indexOf("http://iiif.io/api/presentation/2/context.json") !== -1 || t4["@context"] === "http://www.shared-canvas.org/ns/context.json") || t4["@context"] === "http://iiif.io/api/image/2/context.json" || t4["@id"] && t4["@type"] === "sc:Collection" || t4["@id"] && t4["@type"] === "sc:Manifest" ? (t4["@context"] || (t4["@context"] = "http://iiif.io/api/presentation/2/context.json"), _e.traverseUnknown(t4)) : t4;
  }
  function g2(t4) {
    if ((Array.isArray(t4["@type"]) && t4["@type"].includes("oa:SvgSelector") || t4["@type"] == "oa:SvgSelector") && ("chars" in t4 || "value" in t4))
      return { type: "SvgSelector", value: "chars" in t4 ? t4.chars : t4.value };
    if (t4["@type"] === "oa:FragmentSelector")
      return { type: "FragmentSelector", value: t4.value };
    if (t4["@type"] === "oa:Choice")
      return [g2(t4.default), ...(Array.isArray(t4.item) ? t4.item : [t4.item]).map(g2)];
    if (t4["@type"] == "iiif:ImageApiSelector")
      return { type: "ImageApiSelector", region: "region" in t4 ? t4.region : void 0, rotation: "rotation" in t4 ? t4.rotation : void 0 };
    throw new Error(`Unsupported selector type: ${t4["@type"]}`);
  }

  // node_modules/.pnpm/@iiif+parser@2.2.0/node_modules/@iiif/parser/dist/upgrader.js
  var t3 = be;

  // src/fetch.ts
  function fetchAndUpgrade(input, init) {
    return fetch(input, init).then((resp) => resp.json()).then(t3);
  }
  return __toCommonJS(fetch_exports);
})();
