// @iiif/helpers/dist/vault-actions.js
import {
  SET_META_VALUE,
  SET_META_VALUE_DYNAMIC,
  UNSET_META_VALUE,
  metaActions
} from "./chunk-YAISZCRN.js";
import {
  ADD_MAPPING,
  ADD_MAPPINGS,
  ADD_METADATA,
  ADD_REFERENCE,
  BATCH_ACTIONS,
  BATCH_IMPORT,
  IMPORT_ENTITIES,
  MODIFY_ENTITY_FIELD,
  REMOVE_METADATA,
  REMOVE_REFERENCE,
  REORDER_ENTITY_FIELD,
  REORDER_METADATA,
  REQUEST_COMPLETE,
  REQUEST_ERROR,
  REQUEST_MISMATCH,
  REQUEST_OFFLINE_RESOURCE,
  REQUEST_RESOURCE,
  RESOURCE_ERROR,
  RESOURCE_LOADING,
  RESOURCE_READY,
  UPDATE_METADATA,
  UPDATE_REFERENCE,
  addMapping,
  addMappings,
  addMetadata,
  addReference,
  batchActions,
  batchImport,
  entityActions,
  importEntities,
  mappingActions,
  modifyEntityField,
  removeMetadata,
  removeReference,
  reorderEntityField,
  reorderMetadata,
  requestActions,
  requestComplete,
  requestError,
  requestMismatch,
  requestOfflineResource,
  requestResource,
  updateMetadata,
  updateReference
} from "./chunk-F2Q2OM3J.js";
import "./chunk-G7A32JAG.js";
export {
  ADD_MAPPING,
  ADD_MAPPINGS,
  ADD_METADATA,
  ADD_REFERENCE,
  BATCH_ACTIONS,
  BATCH_IMPORT,
  IMPORT_ENTITIES,
  MODIFY_ENTITY_FIELD,
  REMOVE_METADATA,
  REMOVE_REFERENCE,
  REORDER_ENTITY_FIELD,
  REORDER_METADATA,
  REQUEST_COMPLETE,
  REQUEST_ERROR,
  REQUEST_MISMATCH,
  REQUEST_OFFLINE_RESOURCE,
  REQUEST_RESOURCE,
  RESOURCE_ERROR,
  RESOURCE_LOADING,
  RESOURCE_READY,
  SET_META_VALUE,
  SET_META_VALUE_DYNAMIC,
  UNSET_META_VALUE,
  UPDATE_METADATA,
  UPDATE_REFERENCE,
  addMapping,
  addMappings,
  addMetadata,
  addReference,
  batchActions,
  batchImport,
  entityActions,
  importEntities,
  mappingActions,
  metaActions,
  modifyEntityField,
  removeMetadata,
  removeReference,
  reorderEntityField,
  reorderMetadata,
  requestActions,
  requestComplete,
  requestError,
  requestMismatch,
  requestOfflineResource,
  requestResource,
  updateMetadata,
  updateReference
};
